docs/debugging.md

```diff
@@ -0,0 +1,138 @@
+
+## Debugging the interpreter
+
+As the majority of code running in the interpreter is actually the runtime, the subsequent section could also prove useful.
+The interpreter provides a trace of all the instructions that it executes.
+Every option that you can pass to the interpreter, can also be passed to the compiler prepended with `run-`, for example:
+```
+interpreter foo.bc --verbose
+```
+or 
+```
+veronac foo.verona --run --run-verbose
+```
+The `--verbose` option basically provides a trace of the interpreters operations:
+for example:
+```
+veronac --run --run-verbose region101.verona
+```
+produces
+```
+[ 27b]: Entering function Main.main, argc=1 retc=1 locals=9
+[ 28d]: LOAD_DESCRIPTOR r1, 0x1
+[ 293]: COPY r8, r1
+[ 296]: CALL 7, 0x1
+[ 296]:  Calling function Main.test1, base=r8 argc=1 retc=1 locals=37
+[ 12c]:  LOAD_DESCRIPTOR r1, 0x3
+...
+```
+
+Both the interpreter, and the compiler's run option, can also provide systematic testing for concurrency:
+```
+  veronac-sys --run testsuite/demo/run-pass/dining_phil.verona --run-seed 100 --run-seed_upper 200
+```
+runs 100 seeds sequentially, but testing various interleavings of the runtime. The seeds are replayable.
+
+## Debugging the runtime
+
+The runtime provides two key features to aid debugging:
+
+* Systematic testing
+* Crash logging (A flight recorder that provides the last few seconds of logging)
+
+Both of these can produce high detail logging of the runtime operations.
+
+It is considerably easier to debug a crash in systematic testing, than in standard concurrent execution.
+The majority of tests of the runtime cover a large number of pseudo-random traces, so that the failures can easily be replayed.  The code in
+
+* [src/rt/test/harness.h/SystematicTestingHarness constructor](https://github.com/microsoft/verona/blob/82a00c9d7e0e4f12b08e2b829a599a9ef94c1402/src/rt/test/harness.h#L38-L70)
+* [src/rt/test/harness.h/SystematicTestingHarness::run](https://github.com/microsoft/verona/blob/82a00c9d7e0e4f12b08e2b829a599a9ef94c1402/src/rt/test/harness.h#L73-L114)
+
+highlights all of the features for running with the various logging and random seed approaches.
+
+To compile the run-time with systematic testing you can pass CMake `-DUSE_SYSTEMATIC_TESTING=On`, 
+and then in your program call
+```C++
+Systematic::enable_logging();
+```
+to print the logging, and
+```C++
+Scheduler::get().set_seed(seed);
+```
+specifies the seed for the pseudo-random interleaving.
+The seed to interleaving should be consistent across platforms.
+
+To enable crash logging you can pass CMake `-DUSE_CRASH_LOGGING=On`
+and in your program call
+```C++
+Systematic::enable_crash_logging();
+```
+to enable its output.
+
+Some build artifacts automatically have these settings enabled by default. 
+
+The systematic testing log has the earliest entry first, while the crash logging is in reverse, and contains timestamps.
```

Comment:
I think this should be emphasized somehow, because it's really easy to get wrong and then get really confused.

Actually, as a separate change, maybe the flight recorder dump could start with a note that the log is in reverse?

Reply:
Good idea, I'll add a message at the top of the log.